src/WebForms/VirtualPath.cs

```diff
@@ -3,14 +3,14 @@
 using System.Text;
 using Microsoft.Extensions.DependencyInjection;
 using Microsoft.Extensions.FileProviders;
-using Microsoft.Extensions.Hosting;
+using Microsoft.Extensions.Options;
 using WebForms;
 
 namespace System.Web;
 
 internal sealed class VirtualPath
 {
-    private static IFileProvider Files => HttpRuntimeHelper.Services.GetRequiredService<IHostEnvironment>().ContentRootFileProvider;
+    internal static IFileProvider Files => HttpRuntimeHelper.Services.GetRequiredService<IOptions<WebFormsOptions>>().Value.WebFormsFileProvider;
```

Comment:
If this stays as ContentRootFileProvider, I think, dynamic compilation will find the aspx in the custom folder but VirtualPath will not resolve it because it will expect the path to be /custompath/abc.aspx. Will try again to verify.

Reply:
hmm yeah, I think you're right. That's the challenge with the old aspnet code (and why it was rewritten).... too many static things that can't easily be overridden easily.

Maybe need to add a property to initialize the parsers (where it calls into the VirtualPath) to have the IFileProvider and an overload for VirtualPath that takes a custom IFileProvider.

Reply:
@twsouthwick I am still trying to connect the dots on this. How do you see it manipulating VirtualPath once it's in PageCompilationOptions?

To get the setting, I was thinking of changing this in WebFormsCompilerExtensions.cs:

```cs
private static void AddWebFormsCompilationCore(this IServiceCollection services, Action<PageCompilationOptions> configure)
    {
        services.AddOptions<PageCompilationOptions>()
            .Configure(options =>
            {
                options.WebFormsFileProvider = HttpRuntimeHelper.Services.GetRequiredService<IOptions<WebFormsOptions>>().Value.WebFormsFileProvider;
                options.AddParser<PageParser>(".aspx");
                options.AddParser<MasterPageParser>(".Master");
```

Reply:
I haven't really connected all the dots - I just know that the template parser will need to use this new virtual root.

You're approach is close, but the HttpRuntimeHelper.Services should only be used where DI cannot be used (it's effectively the service locator pattern which is an alternate to DI.... it's there because webforms itself was built without DI and sometimes we need it).

Here, you can use an overload of the configure method for something like:

```csharp
  .Configure<IOptions<WebFormsOptions>>((webFormsOptions, options) =>
{

}
```

then pass in the value to the `AddParser` method so that it can be set on the template parser when it is created.

You could also, I guess for simplicity, use the `HttpRUntimeHelper.Services` within the template parser to grab the value as well.

The main thing, as I mentioned above, is that with this new virtual root, the template parser needs to be aware of it *somehow*.... I'm open to ideas there, but these are the ones I'd start with.

Reply:
@twsouthwick can you point me to the code that is the starting point for dynamically compiled pages. For example, if I type Foo.aspx (with dynamic compilation enabled) which is the earliest part where I can put  breakpoint and traverse the code.

So far going through the code, it looks like all pages are compiled at startup and then a watcher watches for any file changes for recompilation. I looked through all the parsers and they all appear to be part of the compilation that happens at startup.

Need to find how routing to an aspx leads to the dynamically compiled dll. Thanks!

Reply:
Take a look at the design doc here: https://github.com/CoreWebForms/CoreWebForms/blob/main/docs/designs/Handlers.md. Pages are IHttpHandlers and I enabled a general IHttpHandler infrastructure.  DynamicSystemWebCompilation implements IHttpHandlerCollection which is used to generate dynamic endpoints in the ASP.NET Core routing system.

Does that help?